### DISCUSSION TITLE
Require specific dependencies in a group to have a pending update before creating the group PR

### DISCUSSION THREAD

**Discussion by lluiscab** [renovatebot/renovate#26103] [Suggest an Idea]
### Tell us more.

Consider the following configuration:

```json
"packageRules": [
  {
    "matchPackageNames": [
      "library-a",
      "library-b",
      (...)
    ],
    "groupName": "example group",
  }
]
```

## Current 

Right now, this group will be created whenever there's a pending update for any of the dependencies listed in `matchPackageNames`.

## Proposal

I'd like to be able to control when this group PR will be created by *requiring* a subgroup of dependencies to have a pending update in such a way that in order fot the PR to be created, all the dependencies marked as required will need to be present in such PR.

With the configuration above, let's say I only wanted to update the depenendencies in the group if `library-a` was updated. Basically I'd like to never update `library-b` alone and always have it update in "sync" / at the same time as `library-a` is updated. 

Something like the following:

```json
"packageRules": [
  {
    "matchPackageNames": [
      "library-a",
      "library-b"
    ],
    "requiredPackageNames": [
      "library-a"
    ],
    "groupName": "example group",
  }
]
```

When renovate runs, if it only finds an update for `library-a` (matches all the required dependencies), then raise a PR
If it only finds an update for `library-b` (optional, no required dependencies found), then don't raise a PR
If if finds an update for both `library-a` and `library-b` (found all the required dependencies, and some optional as well), then raise a PR.

**Comment by rarkins** [comment]
I think this could be done reasonably elegantly as long as the result is to simply suppress/postpone the PR and not to _ungroup_ them.

I thought about whether this should be a package pattern instead of names for more flexibility but that would also add more complexity. e.g. with a list of names we can know "it must have these 2 packages" whereas with a pattern we'd need to introduce yet another field.

An edge case to consider is what to do if the branch already exists, for whatever reason (e.g. was created prior to the config rule being added, or it was manually approved via dashboard). In that case I think it's best to leave the branch open, otherwise it would never be possible to manually approve.

Perhaps treating this as a new status check (e.g. like `minimumReleaseAge` is) would work.

**Comment by lluiscab** [comment]
I'm not too familiar with the inner workings of Renovate itself, but I think this could very much be an internal status check that just postpones the creation of the MR until all the required packages have a pending update (while maybe also allowing the user to manually create the PR from the dependency dashboard).

In case of an already existing MR, I also think keeping it open is the way to go.

Regarding using package patterns, while that might be helpful, we'd need another parameter to know the minimum amount of matches to find before creating the PR and I think that'd be going a bit too far. For the use case that I'm interested in, being able to specify a list of required packages is more than enough.

Is this something that I might be able to contribute? I might need some pointers on where to start if that's the case.

**Reply by rarkins** [reply]
Yes, definitely! I'd describe this as mid complexity. It wouldn't be that much code, but working out where to put it would be challenging part. It would be most similar to minimumReleaseAge however it would need to be evaluated a _after_ grouping is done in `branchifyUpgrades()`

**Comment by aaronadamsCA** [comment]
I bet this could help keep the Playwright version aligned between Docker and NPM (#23889).

Since Playwright's Docker image is from the Microsoft container registry, it doesn't support `minimumReleaseAge` (#34723).

```
DEBUG: No currentVersionTimestamp for mcr.microsoft.com/playwright
```

Presumably with this feature, the Docker update could be held back until its corresponding NPM update passed its `minimumReleaseAge` check.

**Reply by aaronadamsCA** [reply]
I guess this feature still wouldn't prevent a version mismatch within a group:

<img width="746" alt="image" src="https://github.com/user-attachments/assets/89674e2b-54bb-4b26-89ee-b1209bf0ea89" />

But that could always be another feature to pair nicely with this one.
